### PR TITLE
[Merged by Bors] - feat: port Data.LazyList.Basic

### PR DESCRIPTION
**Pull request by int-y1**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/LazyList/Basic.lean [hunk 1]
```diff
@@ -159,63 +161,84 @@ instance : Monad LazyList where
...
 
+-- Porting note: Rewrote proof of `append_bind`.
 theorem append_bind {α β} (xs : LazyList α) (ys : Thunk (LazyList α)) (f : α → LazyList β) :
-    (@LazyList.append _ xs ys).bind f = (xs.bind f).append ((ys ()).bind f) := by
-  induction xs <;> simp [LazyList.bind, append, *, append_assoc, append, LazyList.bind]
+    (xs.append ys).bind f = (xs.bind f).append (ys.get.bind f) := by
+  match xs with
+  | LazyList.nil => rfl
+  | LazyList.cons x xs =>
+    simp [LazyList.bind, LazyList.append, Thunk.get]
```
Comment (jcommelin) [inline Mathlib/Data/LazyList/Basic.lean:184]: Please replace these non-terminal simps with the output of `simp?`.
Also elsewhere.
(See https://leanprover-community.github.io/glossary.html#non-terminal-simp)

**Review by jcommelin — COMMENTED** [review]
Thanks for helping with the porting effort!

File: Mathlib/Data/LazyList/Basic.lean (hunk 1 above)
Comment (int-y1) [reply Mathlib/Data/LazyList/Basic.lean]: Updated. All non-terminal simps have `only`.
As a side note, `simp?` was slightly broken, so I did some manual work to fix a few simp.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**